pkg/kclient/all.go

```diff
@@ -49,13 +54,15 @@ func getAllResources(client dynamic.Interface, apis []apiResource, ns string, se
 				errResult = err
 				return
 			}
-			mu.Lock()
-			out = append(out, v...)
-			mu.Unlock()
+			outChan <- v
 			klog.V(4).Infof("[query api]  done: %s, found %d apis", a.GroupVersionResource(), len(v))
 		}(api)
 	}
 
+	for i := 0; i < len(apisOfInterest); i++ {
+		out = append(out, <-outChan...)
```

Comment:
If one of the goroutines fired returns an error, it will not write to `outChan`. So attempting to read from `outChan` `len(apisOfInterest)` times could block and hang up the `odo logs` command, I think.

Reply:
> If one of the goroutines fired returns an error, it will not write to `outChan`. So attempting to read from `outChan` `len(apisOfInterest)` times could block the `odo logs` command, I think.

Confirmed by simulating an error from time to time with the following diff:
```diff
diff --git a/pkg/kclient/all.go b/pkg/kclient/all.go
index f1a0abe52..85bdbe9ff 100644
--- a/pkg/kclient/all.go
+++ b/pkg/kclient/all.go
@@ -43,12 +43,16 @@ func getAllResources(client dynamic.Interface, apis []apiResource, ns string, se
        klog.V(2).Infof("starting to concurrently query %d APIs", len(apis))
 
        var errResult error
+       var i int
        for _, api := range apisOfInterest {
                wg.Add(1)
-               go func(a apiResource) {
+               go func(a apiResource, j int) {
                        defer wg.Done()
                        klog.V(4).Infof("[query api] start: %s", a.GroupVersionResource())
                        v, err := queryAPI(client, a, ns, selector)
+                       if err == nil && j%10 == 5 {
+                               err = fmt.Errorf("XXX an error for j=%d", j)
+                       }
                        if err != nil {
                                klog.V(4).Infof("[query api] error querying: %s, error=%v", a.GroupVersionResource(), err)
                                errResult = err
@@ -56,7 +60,8 @@ func getAllResources(client dynamic.Interface, apis []apiResource, ns string, se
                        }
                        outChan <- v
                        klog.V(4).Infof("[query api]  done: %s, found %d apis", a.GroupVersionResource(), len(v))
-               }(api)
+               }(api, i)
+               i++
        }
 
        for i := 0; i < len(apisOfInterest); i++ {

```

Reply:
f3649ef

Reply:
Good catch! Initially I tried by `close`'ing the channel when a go routine encountered an error. But it was prone to panic and race conditions. Eventually, I ended up implementing the pattern of `range`'ing over a channel in a go routine. Take a look.